### PR TITLE
feat: add pallet-contracts

### PR DESCRIPTION
**Pull request by peterwht**

### CODE REVIEW

File: runtime/src/contracts_config.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+pub enum AllowBalancesCall {}
+
+impl frame_support::traits::Contains<RuntimeCall> for AllowBalancesCall {
+	fn contains(call: &RuntimeCall) -> bool {
+		matches!(call, RuntimeCall::Balances(BalancesCall::transfer_allow_death { .. }))
+	}
+}
+
+// Unit = the base number of indivisible units for balances
+const UNIT: Balance = 1_000_000_000_000;
```
Comment (evilrobot-01) [inline runtime/src/contracts_config.rs:20]: These are defined in the runtime?

Comment (peterwht) [reply runtime/src/contracts_config.rs]: Good catch. Fixed

File: runtime/src/contracts_config.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		(Default::default(), Default::default())
+	}
+}
+
+parameter_types! {
+	pub const DepositPerItem: Balance = deposit(1, 0);
+	pub const DepositPerByte: Balance = deposit(0, 1);
+	pub Schedule: pallet_contracts::Schedule<Runtime> = schedule::<Runtime>();
+	pub const DefaultDepositLimit: Balance = deposit(1024, 1024 * 1024);
+	pub const CodeHashLockupDepositPercent: Perbill = Perbill::from_percent(0);
```
Comment (evilrobot-01) [inline runtime/src/contracts_config.rs]: Trappist has this configured as 30%: https://github.com/paritytech/trappist/blob/a07b7e84224c53d16dbd1d5e3fb8797f33d12723/runtime/trappist/src/contracts.rs#L38

File: runtime/src/contracts_config.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	/// The safest default is to allow no calls at all.
+	///
+	/// Runtimes should whitelist dispatchables that are allowed to be called from contracts
+	/// and make sure they are stable. Dispatchables exposed to contracts are not allowed to
+	/// change because that would break already deployed contracts. The `RuntimeCall` structure
+	/// itself is not allowed to change the indices of existing pallets, too.
+	type CallFilter = AllowBalancesCall;
+	type DepositPerItem = DepositPerItem;
+	type DepositPerByte = DepositPerByte;
+	type CallStack = [pallet_contracts::Frame<Self>; 23];
```
Comment (evilrobot-01) [inline runtime/src/contracts_config.rs]: Trappist has this configured as 5: https://github.com/paritytech/trappist/blob/a07b7e84224c53d16dbd1d5e3fb8797f33d12723/runtime/trappist/src/contracts.rs#L58

File: runtime/src/contracts_config.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    /// The safest default is to allow no calls at all.
+    ///
+    /// Runtimes should whitelist dispatchables that are allowed to be called from contracts
+    /// and make sure they are stable. Dispatchables exposed to contracts are not allowed to
+    /// change because that would break already deployed contracts. The `RuntimeCall` structure
+    /// itself is not allowed to change the indices of existing pallets, too.
+    type CallFilter = AllowBalancesCall;
+    type DepositPerItem = DepositPerItem;
+    type DepositPerByte = DepositPerByte;
+    type CallStack = [pallet_contracts::Frame<Self>; 23];
```
Comment (evilrobot-01) [inline runtime/src/contracts_config.rs:70]: Note: probably doesnt matter for now, but Trappist has this configured as 5: https://github.com/paritytech/trappist/blob/a07b7e84224c53d16dbd1d5e3fb8797f33d12723/runtime/trappist/src/contracts.rs#L58

File: runtime/src/contracts_config.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        (Default::default(), Default::default())
+    }
+}
+
+parameter_types! {
+    pub const DepositPerItem: Balance = deposit(1, 0);
+    pub const DepositPerByte: Balance = deposit(0, 1);
+    pub Schedule: pallet_contracts::Schedule<Runtime> = schedule::<Runtime>();
+    pub const DefaultDepositLimit: Balance = deposit(1024, 1024 * 1024);
+    pub const CodeHashLockupDepositPercent: Perbill = Perbill::from_percent(0);
```
Comment (evilrobot-01) [inline runtime/src/contracts_config.rs:51]: Note: probably doesnt matter for now, but Trappist has this configured as 30%: https://github.com/paritytech/trappist/blob/a07b7e84224c53d16dbd1d5e3fb8797f33d12723/runtime/trappist/src/contracts.rs#L38

**Review by evilrobot-01 — APPROVED** [review]

**Pull request merged**